Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -0,0 +1,306 @@
+          return *this;
+        } else {
+          replacing_entry = true;
+        }
+      }
+    }
+
+    // Copy the segment before the found position. If not found, this is
+    // everything.
+    auto copy = std::make_shared<array_type>();
+    copy->append(begin(), pos);
+
+    // Copy the value to be inserted.
+    copy->append(std::move(pair));
+
+    if (replacing_entry) {
+      // Skip the thing at pos because it compares the same as the pair above.
+      ++pos;
+    } else {
+      // If inserting at the end or the key at pos is not equal to what we're
+      // inserting, then increase the size.
+      assert(size() < kFixedSize);
+    }
+
+    // Copy everything after pos (if anything).
+    copy->append(pos, current_end);
+    return wrap(copy);
+  }
+
+  /**
+   * Creates a new map identical to this one, but with a key removed from it.
+   *
+   * @param key The key to remove.
+   * @return A new dictionary without that value.
+   */
+  this_type erase(const K& key) const {
+    const_iterator current_end = end();
+    const_iterator pos = find(key);
+    if (pos == current_end) {
+      return *this;
+    } else {
+      auto copy = std::make_shared<array_type>();
+      copy->size_ = size() - 1;
+      auto copy_end = std::copy(begin(), pos, copy->begin());
+      if (pos + 1 < current_end) {
+        std::copy(pos + 1, end(), copy_end);
+      }
+      return wrap(copy);
+    }
+  }
+
+  /**
+   * Finds a value in the map.
+   *
+   * @param key The key to look up.
+   * @return An iterator pointing to the entry containing the key, or end() if
+   *     not found.
+   */
+  const_iterator find(const K& key) const {
+    const_iterator not_found = end();
+    const_iterator lower_bound = LowerBound(key);
+    if (lower_bound != not_found && !key_comparator_(key, *lower_bound)) {
+      return lower_bound;
+    } else {
+      return not_found;
+    }
+  }
+  // indexof
+
+  /** Returns true if the map contains no elements. */
+  bool empty() const {
+    return size() == 0;
+  }
+
+  /** Returns the number of items in this map. */
+  size_type size() const {
+    return array_->size_;
+  }
+
+  /**
+   * Returns an iterator pointing to the first entry in the map. If there are
+   * no entries in the map, begin() == end().
+   */
+  const_iterator begin() const {
+    return array_->begin();
+  }
+
+  /**
+   * Returns an iterator pointing past the last entry in the map.
+   */
+  const_iterator end() const {
+    return array_->end();
+  }
+
+ private:
+  static array_pointer EmptyArray() {
+    static const array_pointer kEmptyArray =
+        std::make_shared<const array_type>();
+    return kEmptyArray;
+  }
+
+  ArraySortedMap(const array_pointer& array,
+                 const key_comparator_type& key_comparator) noexcept
+      : array_(array), key_comparator_(key_comparator) {
+  }
+
+  this_type wrap(const array_pointer& array) const noexcept {
+    return this_type(array, key_comparator_);
```

Comment:
If array.size() == 0, perhaps this should return `this_type{EmptyArray(), key_comparator_}`? IIUC, currently all default-constructed maps will share EmptyArray, but not e.g. this map:

```
ArraySortedMap map{{1,1}};
map.remove(1);
```